gateleen-core/src/main/java/org/swisspush/gateleen/core/util/HttpHeaderUtil.java

```diff
@@ -61,18 +68,43 @@
      * Helper method to check the presence of a http header matching the provided pattern. Every header entry is checked
      * in the format KEY: VALUE
      *
-     * @param headers   The Map with the header key - value pairs to be evaluated
+     * @param headers        The Map with the header key - value pairs to be evaluated
      * @param headersPattern The pattern to match the headers against.
      * @return True if a matching header was found, false otherwise
      */
     public static <T extends MultiMap> boolean hasMatchingHeader(@Nonnull T headers, @Nonnull Pattern headersPattern) {
         for (Map.Entry<String, String> entry : headers) {
             String header = entry.getKey() + ": " + entry.getValue();
-            if(headersPattern.matcher(header).matches()){
+            if (headersPattern.matcher(header).matches()) {
                 return true;
             }
         }
         return false;
     }
 
+    /**
+     * Merges headers, makes sure that only one value of header {@link HttpRequestHeader#CONTENT_LENGTH} ends up in the result.
+     */
+    public static void mergeHeaders(@Nonnull MultiMap destination, @Nonnull MultiMap source, @Nullable String context) {
+        String destinationContentLength = "";
+        String sourceContentLength = "";
+
+        if (HttpRequestHeader.containsHeader(destination, HttpRequestHeader.CONTENT_LENGTH)) {
+            destinationContentLength = destination.get(HttpRequestHeader.CONTENT_LENGTH.getName());
+        }
+        if (HttpRequestHeader.containsHeader(source, HttpRequestHeader.CONTENT_LENGTH)) {
+            sourceContentLength = source.get(HttpRequestHeader.CONTENT_LENGTH.getName());
+        }
+        if (destinationContentLength.isEmpty() || sourceContentLength.isEmpty()) {
+            destination.addAll(source);
+        } else {
+            if (!destinationContentLength.equals(sourceContentLength)) {
+                LOG.warn("Content-Length values do not match {} != {} for request to url {}",
+                        destinationContentLength, sourceContentLength, context);
```

Comment:
Not sure if this is a good idea to just continue in this case. I think we should fail if the values differ (eg throw or whatever).

Continuing here potentially leads to unhealthy out-of-sync http connections.

Reply:
Basically agree. It would mean to implement the error handling for some 10 usages of this merge method. Hmm, not sure if we can justify the effort. But I changed the log into an error to indicate that there is clearly something not as it should be.